Orphan bead handling wrt DB

Sansh2356 opened the pull request:
- Currently the `orphan` beads were not being persisted in DB leading to foreign key violation and causing error in subsequent bead propagation .
- This PR adds the functionality for connecting the `Braid` with the `db_handler` using `db_tx` and thus persisting the beads if any present in `orphan buffer` as soon as all of its parents have been received . 
- It also removes request in `BeadSync` for the beads whose parents have not yet been received i.e. it is orphan bead which was redundant in nature causing unnecessary utilization of bandwidth .
- Changes verbosity of logs and updating the tests accordingly .

The above was initial though but over the discussions me @mcelrath arrived at a conclusion that tight coupling b/w DB and `Braid` is not required and must be kept as separated as possible .

- Removal of locks passed on for `Braid` into main thread only reducing the causes as well as checking of potential race conditions arising eventually leading to deadlocks. 
- Orphan persistence using the simple length check instead of anything else .
- Removal of `Timestamp` based checking for `IBD` for the purpose of connecting DAG after receiving beads from sync nodes as explained under #306 and `ibd_manager` moving into `peer_manager`.
- The mining condition is to be moved into `stratum` component instead depending dynamically over `Orphan-Occupancy` above certain threshold `alpha` defined as latency param which will eventually done in forth-coming PR  and adding batched insertions during IBD.
- This also adds IBD retry to be initiated per request basis for respective timeouts to switch between sync peers to continue IBD from the previous state.
- Adding ibd_flag for preventing downstream connections.
- Replacing the `Arc<Mutex<T>>` wrapping around db connection pool to avoid unnecessary locking along with the clearing IBD buffers in case of stale or dropped and retried requests and clearing buffer after ibd is completed.
- It also adds batched insertions along with sequential insertions for both orphan beads as well as the parent  beads.

Sansh2356 committed:
fix: Persisting orphan beads in DB and removing redundant request calls in BeadSync

Sansh2356 committed:
fix: Persisting orphan beads in DB and removing redundant request calls in BeadSync

mcelrath reviewed (changes requested):
This PR creates a circular dependency, Braid <-> DBHandler.

In main.rs:517 around `AddBeadStatus::BeadAdded` we are writing new beads to the DB on a successful extend (which will automatically also add orphans to the braid if possible). So all we really need to do is update main.rs to save not only the new bead, but instead all new beads in the braid since the last save. 

For instance, Before calling `braid_data.extend(&bead)` at main.rs:477, save `let braid_prev_len = braid_data.beads.len()` and then when writing to the DB, write all beads between `braid_prev_len` and the new `braid_data.beads.len()`.

Sansh2356 committed:
refactor: removing braid atomic refs for locking in main thread

Sansh2356 committed:
refactor: removing braid atomic refs for locking in main thread

Sansh2356 committed:
feat: orphan beads persisted without tight coupling

Sansh2356 committed:
feat: orphan beads persisted without tight coupling

Sansh2356 committed:
removing ibd_manager and timestamp based checking in favour of sync checks into peer manager module

changing insert bead function signature

changing verbosity of ibd logs

Sansh2356 committed:
removing ibd_manager and timestamp based checking in favour of sync checks into peer manager module

changing insert bead function signature

changing verbosity of ibd logs

mcelrath commented on node/src/db/db_handlers.rs:
```diff
@@ -208,69 +344,6 @@ impl DBHandler {
         }
     }
 }
-pub fn prepare_bead_tuple_data(
```
This function has moved in the file, making comparison difficult. Unless it's really important that it appear at this place in the file, please move your modified implementation back to this location.

mcelrath commented on node/src/db/mod.rs:
```diff
@@ -8,10 +10,9 @@ pub mod init_db;
-        relative_json: String,
-        parent_timestamp_json: String,
-        bead_id: usize,
+        curr_beads: Vec<Bead>,
```
Why do we need `bead_to_insert`, `curr_beads` and `removed_orphans` too? Why not just one `Vec<Bead>`?

Sansh2356 replied:
`bead_to_insert` is required for bead_id as it is not necessary that bead_id will always be equal to last_idx of  `Vec<Bead>` of `Braid`, we can directly pass on `bead_id` instead .

mcelrath commented on node/src/db/mod.rs:
```diff
@@ -8,10 +10,9 @@ pub mod init_db;
-        parent_timestamp_json: String,
-        bead_id: usize,
+        curr_beads: Vec<Bead>,
+        bead_index_mapping: HashMap<bitcoin::BlockHash, usize>,
```
Please use the `BeadHash` and `BeadIdx` types.

mcelrath reviewed (changes requested):
Overall, this seems overly complex to me. I don't understand why `prepare_bead_tuple_data` takes so many arguments, its job is just to make json tuples for insert. Why does it take so many arguments?

Likewise InsertBeadSequentially should probably just be called `InsertBeads` and it should take a `Vec<Bead>`. Where they need to get things like BeadIdx, why just not pass in a reference to the Braid?

zaidmstrr commented on node/src/peer_manager/mod.rs:
```diff
@@ -26,6 +32,17 @@ pub struct PeerInfo {
     ip_addr: Option<IpAddr>,
+    /// Tracks the most recent tips received from each peer during IBD.
+    /// (peer_id --> Vec<BeadHash>)
+    tips_mapping: Vec<BeadHash>,
```
The names here are confusing and don't represent the true nature of the variable use. Same for `batch_mapping`, `get_bead_mapping`, choose a name which truly states that these variables are used for IBD and for the specific behaviour.

Sansh2356 replied:
Will do the changes .

zaidmstrr commented on node/src/peer_manager/mod.rs:
```diff
@@ -1,8 +1,14 @@
 
+pub const IBD_BATCH_SIZE: usize = 500;
+pub const MAX_IBD_RETRIES: u64 = 10;
+pub const IBD_TRIGGER_AFTER: u64 = 20;
```
Why are we waiting for 20 sec in `IBD_TRIGGER_AFTER`? I think instead of waiting for a fixed period of time we can use some ready signal which will allow us to trigger the IBD. Example signal ready after peer discovery, etc.

Sansh2356 replied:
Yeah, we can do minimum connected peer and stall until that threshold is not achieved .

zaidmstrr commented on node/src/db/db_handlers.rs:
```diff
@@ -169,12 +214,25 @@ impl DBHandler {
                 BraidpoolDBTypes::InsertTupleTypes { query } => match query {
                     InsertTupleTypes::InsertBeadSequentially {
```
Here the beads are added sequentially to the persistence db even when we are fetching the beads in batches during IBD. What if we add another type which allows us to add the beads in batches to the db? We can do this by first validating all the beads using `braid.extend()` and then using a single transaction for the db to add all the beads at once. And elsewhere (without IBD), sequential bead insert is sufficient.

zaidmstrr reviewed (commented):
I reviewed this PR and found some areas of improvement and gave some suggestions. You are free to discuss them below.

Sansh2356 committed:
adding batched insertions along with reducing insertion logic redundancy

Sansh2356 committed:
adding batched insertions along with reducing insertion logic redundancy

Sansh2356 committed:
separating raw batch-query to sub-queries and adding test for batch

Sansh2356 committed:
separating raw batch-query to sub-queries and adding test for batch

Sansh2356 committed:
removing original master head to adapt current branch changes during rebase

removing typo

removing not needed file

Sansh2356 committed:
removing original master head to adapt current branch changes during rebase

removing typo

removing not needed file

Sansh2356 committed:
removing GetParents call immediately after orphan_bead is received

Sansh2356 committed:
removing GetParents call immediately after orphan_bead is received

Sansh2356 committed:
changing verbosity of logs and formation

Sansh2356 committed:
changing verbosity of logs and formation

Sansh2356 committed:
adding more error types and removing unwanted unwraps

zaidmstrr commented on node/src/db/db_handlers.rs:
```diff
@@ -126,39 +219,399 @@ impl DBHandler {
+        };
+
+        let mut inserted_count = 0u32;
+        let beads_ref = beads.clone();
```
The variable `beads_ref` is computationally expensive with no real benefits and needs to be completely removed from the database. The only purpose it serves is logging.

zaidmstrr commented on node/src/db/db_handlers.rs:
```diff
@@ -126,39 +219,399 @@ impl DBHandler {
+                        if let Ok(parent_ts) =
+                            serde_json::from_str::<Vec<serde_json::Value>>(&parent_ts_json)
+                        {
+                            // warn!("PARENTS JSON STRING - {:?}",parent_ts);
```
This line needs to be removed.

zaidmstrr commented on node/src/db/db_handlers.rs:
```diff
@@ -126,39 +219,399 @@ impl DBHandler {
+    ) -> anyhow::Result<(String, String, String)> {
+        let bead_id = bead_index_mapping
+            .get(&bead.block_header.block_hash())
+            .unwrap()
```
It's good to not use `unwrap` here.

zaidmstrr commented on node/src/db/db_handlers.rs:
```diff
@@ -126,39 +219,399 @@ impl DBHandler {
+        let mut relatives = Vec::new();
+        let mut parent_ts = Vec::new();
+        for parent in &bead.committed_metadata.parents {
+            let (parent_index, parent_timestamp) = bead_index_mapping.get(parent).unwrap();
```
You can also remove the `unwrap` from here.

zaidmstrr commented on node/src/db/db_handlers.rs:
```diff
@@ -126,39 +219,399 @@ impl DBHandler {
+                        original_bead, bead
+                    );
+                    for parent_hash in &original_bead.committed_metadata.parents {
+                        let id = bead_index_mapping.get(parent_hash).unwrap();
```
Same here you can remove `unwrap()`.

Sansh2356 replied:
I have addressed all the unwraps in `docker-test-env` branch already, i will pick that commit here .

zaidmstrr reviewed (commented):
I have reviewed the PR and found some issues. You can look at them and resolve them and discuss them.

Sansh2356 committed:
fix(typo): removing commented logs and fmt check

chore: rebasing over dev resolving conflicts in lockfile

mcelrath commented:
● Blockers (detail)

   1. IBD queue/tips are unbounded and never reset  
    - Where: node/src/peer_manager/mod.rs around handle_update_ibd_* and handle_fetch_ibd_* (lines ~149–188).  
    - What happens: ibd_bead_queue and ibd_peer_tips are only extended and then cloned for fetch; they’re not cleared or
   capped.  
    - Why this is a blocker: A peer can keep sending large batches, causing memory growth and duplicate bead hashes. 
  Retries then re‑request already‑seen ranges, and queues keep growing—classic DoS surface and correctness risk.  
    - Fix direction: Replace vs extend (or drain after fetch), cap at IBD_BATCH_SIZE, and reset per‑peer IBD state on 
  retry/complete.
   2. Batch insert can re‑insert resolved orphans  
    - Where: node/src/main.rs around batch IBD insert (lines ~951–955).  
    - What happens: InsertBeadsBatch is called with all received beads plus removed_orphans. Orphans that were resolved 
  earlier in the batch can appear twice.  
    - Why this is a blocker: DB has UNIQUE/PK on bead hash; duplicates can trigger a rollback and stall IBD.  
    - Fix direction: Only include AddBeadStatus::BeadAdded in the batch list and dedupe against removed_orphans.
   3. Partial tuple prep leaves DB inconsistent  
    - Where: node/src/db/db_handlers.rs around the bulk insert path (lines ~331–371).  
    - What happens: The Bead row is inserted even if prepare_bead_tuple_data fails; related rows 
  (Relatives/ParentTimestamps/Transactions) are skipped.  
    - Why this is a blocker: Leaves orphaned Bead rows without relations—breaks invariants and later queries.  
    - Fix direction: Validate tuple prep first or roll back the bead row; keep insert and tuple creation in a single 
  transaction that fails atomically.
   4. HWP comparator errors masked  
    - Where: node/src/braid/mod.rs around highest_work_path (line ~1068).  
    - What happens: comparator errors are treated as equal, which can make HWP nondeterministic.  
    - Why this is a blocker: HWP is consensus‑critical; nondeterminism risks divergence.  
    - Fix direction: propagate/handle comparator errors explicitly.
   5. IBD pruning order risks skipping beads  
    - Where: node/src/main.rs around IBD pruning/queue creation (line ~1059).  
    - What happens: pruning stops at tips and iterates cohorts in an unordered way; hashes may be truncated and beads 
  skipped.  
    - Why this is a blocker: IBD can miss beads, causing inconsistent DAG state.

  ---------------------------------------------------------------------------------------------------------------------

  Failed tests (detail)

  cargo test -p node → 95 passed, 8 failed.

  AddrInUse failures (port already bound):

   - behaviour::tests::test_bead_request_handling (node/src/behaviour/tests.rs:94)
   - behaviour::tests::test_floodsub_message_propagation (node/src/behaviour/tests.rs:289)
   - rpc_server::test_extend_rpc (node/src/rpc_server.rs:1044)
   - rpc_server::test_get_mining_info_rpc (node/src/rpc_server.rs:1044)
   - rpc_server::test_get_ipc_stats_rpc_simple (node/src/rpc_server.rs:1044)
   - rpc_server::test_get_tips_cli_flow (node/src/rpc_server.rs:1044)
   - rpc_server::test_unstage_transactions_rpc_simple (node/src/rpc_server.rs:1044)

  Meaning: These tests try to bind fixed ports and collided with existing listeners. This is environmental flakiness,
  not necessarily caused by PR code. We can rerun with isolated ports or ensure no lingering services are running.

  Logic failure:

   - stratum::test::server_start_test (node/src/stratum.rs:2119)
  Assertion expected 3, got 6. This looks like a behavioral change from PR #309 (likely more connections/events 
  recorded), and needs code‑level investigation.

Sansh2356 commented:
> ● Blockers (detail)
> 
>     1. IBD queue/tips are unbounded and never reset
>        - Where: node/src/peer_manager/mod.rs around handle_update_ibd_* and handle_fetch_ibd_* (lines ~149–188).
>        - What happens: ibd_bead_queue and ibd_peer_tips are only extended and then cloned for fetch; they’re not cleared or
>        capped.
>        - Why this is a blocker: A peer can keep sending large batches, causing memory growth and duplicate bead hashes.
>        Retries then re‑request already‑seen ranges, and queues keep growing—classic DoS surface and correctness risk.
>        - Fix direction: Replace vs extend (or drain after fetch), cap at IBD_BATCH_SIZE, and reset per‑peer IBD state on
>        retry/complete.
> 
>     2. Batch insert can re‑insert resolved orphans
>        - Where: node/src/main.rs around batch IBD insert (lines ~951–955).
>        - What happens: InsertBeadsBatch is called with all received beads plus removed_orphans. Orphans that were resolved
>        earlier in the batch can appear twice.
>        - Why this is a blocker: DB has UNIQUE/PK on bead hash; duplicates can trigger a rollback and stall IBD.
>        - Fix direction: Only include AddBeadStatus::BeadAdded in the batch list and dedupe against removed_orphans.
> 
>     3. Partial tuple prep leaves DB inconsistent
>        - Where: node/src/db/db_handlers.rs around the bulk insert path (lines ~331–371).
>        - What happens: The Bead row is inserted even if prepare_bead_tuple_data fails; related rows
>        (Relatives/ParentTimestamps/Transactions) are skipped.
>        - Why this is a blocker: Leaves orphaned Bead rows without relations—breaks invariants and later queries.
>        - Fix direction: Validate tuple prep first or roll back the bead row; keep insert and tuple creation in a single
>        transaction that fails atomically.
> 
>     4. HWP comparator errors masked
>        - Where: node/src/braid/mod.rs around highest_work_path (line ~1068).
>        - What happens: comparator errors are treated as equal, which can make HWP nondeterministic.
>        - Why this is a blocker: HWP is consensus‑critical; nondeterminism risks divergence.
>        - Fix direction: propagate/handle comparator errors explicitly.
> 
>     5. IBD pruning order risks skipping beads
>        - Where: node/src/main.rs around IBD pruning/queue creation (line ~1059).
>        - What happens: pruning stops at tips and iterates cohorts in an unordered way; hashes may be truncated and beads
>        skipped.
>        - Why this is a blocker: IBD can miss beads, causing inconsistent DAG state.
> 
> 
> Failed tests (detail)
> 
> cargo test -p node → 95 passed, 8 failed.
> 
> AddrInUse failures (port already bound):
> 
>     * behaviour::tests::test_bead_request_handling (node/src/behaviour/tests.rs:94)
> 
>     * behaviour::tests::test_floodsub_message_propagation (node/src/behaviour/tests.rs:289)
> 
>     * rpc_server::test_extend_rpc (node/src/rpc_server.rs:1044)
> 
>     * rpc_server::test_get_mining_info_rpc (node/src/rpc_server.rs:1044)
> 
>     * rpc_server::test_get_ipc_stats_rpc_simple (node/src/rpc_server.rs:1044)
> 
>     * rpc_server::test_get_tips_cli_flow (node/src/rpc_server.rs:1044)
> 
>     * rpc_server::test_unstage_transactions_rpc_simple (node/src/rpc_server.rs:1044)
> 
> 
> Meaning: These tests try to bind fixed ports and collided with existing listeners. This is environmental flakiness, not necessarily caused by PR code. We can rerun with isolated ports or ensure no lingering services are running.
> 
> Logic failure:
> 
>     * stratum::test::server_start_test (node/src/stratum.rs:2119)
>       Assertion expected 3, got 6. This looks like a behavioral change from PR [Orphan bead handling wrt DB  #309](https://github.com/braidpool/braidpool/pull/309) (likely more connections/events
>       recorded), and needs code‑level investigation.

All the tests are passing can checkout to branch and run `cargo test -p node`. 

Apart from that the provided suggestions 1 i already had mind for draining all the required states after IBD has been completed, apart from that in 3 i agree that `continue` should be replaced with hard-errors leading to db-component being shut-down since any disorder will lead to non-consistent state for every forward state so this will be replace by `DBErrors` instead and 4 the error must be replaced.


Apart from the above 3 remaining 2 are not called for, respectively because the `removed_orphan` is for a single batch only hence cannot occur twice for the same batch and in the other received beads are in order so pruning them according to the tips difference will and should be handled by the `cohort` as well as `extend` functionalities respectively .

Sansh2356 committed:
chore(ibd): Adding ibd peer data to be replaced at retry,replacing continue in db_handlers

Currently the buffers being kept for ibd should be drained before retry and replaced
at each retry and similarly if the ibd gets completed the in-memory data can be dropped.

Removing the continue branches from db_handlers in place of concrete fallible error types
in case any of the persistance mapping transaction is rollbacked or in case of error

preventing inconsistent and unnecessary further iteration of remaining beads in case of
both orphan and non-orphand beads.

Sansh2356 committed:
chore(ibd): Removing Mutex lock around connection pool and query params,logs

This commit removes the Arc<Mutex<T>> locking around Connection Pool
from sqlx since the inner type is already Arc<T> which further is
wrapped at granular level to provide interior mutability thus is
both Send+Sync and it removes the logging loop in batch insertion
which was being used for testing and adding the check for duplicate
bead that are received during the floodsub propagation

Sansh2356 committed:
chore(ibd): Adding sync_flag for IBD completion and re-requesting IBD in case of failure or timeout

This adds a ibd_complete atomic flag share between stratum server in main swarm loop
that will prevent the mining connections from downstream miners until IBD is completed.

It also modifies peer_manager module to add request monitoring in case of timeout
or in-between peer is disconnected .

Sansh2356 committed:
fix: Updating locking scopes and changing serde functions during db insertions

This removes the holding of locks in bead_sync loop and reducing the contention between
locks, along with that it also improves upon the existing bulk insertion method in which
if the threshold of BATCH_SIZE is reached then it chunks into min batch size and insert
each chunk.

It also changes the prototyping of value serialization function in db_handlers.rs
to reduce the round-trip between serde and concrete value and removing unused params of
functions such as in SwarmHandler and removing oneshot for sending intermediate data
to peer manager required for bead_sync.

Sansh2356 committed:
feat(ibd): Adding pagination in beadhash to limit the beadhash memory overhead

We already had batches based bead fetching on the basis of beadhashes, but
the beadhashes were not batched leading to potential memory overhead during
the initial `GetBeadsAfter` call.

In this we limit the call of beadhashes to batch size and changes in `Codec`
implementation to take the buffer instead of usize::MAX to the framesize
to limit the req/resp call.

chore(ibd): Redoing some comments and updating test in db handler

zaidmstrr commented on node/src/peer_manager/mod.rs:
```diff
@@ -91,23 +122,154 @@ impl PeerManager {
+
+    /// Record an in-flight IBD request to `peer_id`. Called immediately after
+    /// `bead_sync.send_request(...)` returns its `OutboundRequestId`.
+    pub fn set_ibd_inflight(&mut self, peer_id: PeerId, request_id: OutboundRequestId) {
```
The function has no guard to hold the correct invariant. Thus, can include the possible race conditions. Also, this is called in many places inside `main.rs` under independent code blocks.

zaidmstrr commented on node/src/braid/mod.rs:
The `orphan_beads` Vec is an unbounded buffer and has no checks on the capacity limits. Any malicious peer can send any number of malformed beads with unknown parents and can crash the buffer with OOM.

zaidmstrr commented on node/src/main.rs:
```diff
@@ -152,28 +150,20 @@ async fn main() -> Result<(), Box<dyn Error>> {
+    let mut ibd_initiated = false;
+
+    // Flag to indicate IBD completion - prevents miner connections until synced
+    let ibd_complete = Arc::new(AtomicBool::new(false));
```
The variable is never set to `false` after being used or setting it to true. Which can introduce unexpected behaviour later in the code.

zaidmstrr reviewed (commented):
I found some possible issues that needs to taken a look. While the reviewing process is still going on from my side. Until then you can look at the issues I have found.

mcelrath commented:
This report summarizes the open findings from the re-review after rebasing onto the updated upstream.

1. **Security | High | `node/src/main.rs:901` – Unbounded `GetAllBeads` response.** Any peer can request the full DAG, which clones and serializes all beads. **Impact:** peer-triggered CPU/RAM exhaustion and outbound bandwidth spikes. **Fix:** remove or gate `GetAllBeads`, use paged `GetBeadsAfter` only, and enforce a hard response-size cap before encoding.

Bob's take: valid. Use `GetBeadsAfter` only with paging. Add a config setting which limits the number of simultaneous peers we will serve IBD to. (Bitcoin uses 8 "full relay" peers)

2. **Database | High | `node/src/main.rs:1007` – Duplicate inserts in batch.** `removed_orphans` and `beads_to_insert` are concatenated without de-duplication, so resolved orphans can appear twice. **Impact:** Bead PK/UNIQUE violations and full batch rollback, stalling IBD. **Fix:** de-dup by `BeadHash` before batching (preferred) or use `INSERT OR IGNORE` with guarded dependent inserts.

Bob's take: valid. IMHO the whole flow is overly complex. You have three data structures representing "beads to be added" (`beads_to_insert`, `removed_orphans`, and InsertBeadSequentially from floodsub). Ideally there should be just ONE interface, `InsertBeads(Vec<Bead>)` Also consider using an automatically de-duped data structure like HashSet or HashMap.

3. **Performance | Critical | `node/src/main.rs:607/1007` – Orphan-tail cloning under braid write lock.** Per-bead cloning of orphan tail and bead-index mapping occurs under the write lock, producing O(n^2) work. **Impact:** long lock holds and latency spikes as DAG grows. **Fix:** avoid tail cloning per bead; have braid return an explicit resolved-orphans set and process outside the write lock.

Bob's take: valid. The simplest fix to this is to have `extend()` return the list of beads added **including** any resolved orphans, rather than trying to guess about orphans after the fact. (main.rs:589 and main.rs:988)

4. **Performance | High | `node/src/braid/mod.rs:197` – O(n^2) orphan processing.** `process_orphan_beads` removes from a Vec and recurses, repeatedly scanning/shifting. **Impact:** quadratic time on large orphan backlogs. **Fix:** use a queue (`VecDeque`) plus lookup (`HashSet`/index map), or an iterative retain/partition pass.

Bob's take: This is #299 territory, those algorithmic improvements need to be a separate PR and should be fixed there.

5. **Performance | Medium | `node/src/db/db_handlers.rs:539` – N+1 orphan reinserts.** `InsertBeadSequentially` reinserts removed orphans one-by-one with serialization and DB I/O. **Impact:** sync lag for large orphan tails. **Fix:** batch orphan reinserts using the bulk insert path.

Bob's take: valid. Build a vector and use `InsertBeadsBatch` instead. Sequential inserts in SQL is very slow compared to batch inserts.

6. **Rust | High | `node/src/main.rs:282` – `expect` on behaviour creation.** `BraidPoolBehaviour::new(...).expect(...)` panics on startup failures. **Impact:** non-recoverable crash instead of error propagation. **Fix:** return/propagate the error (e.g., map to `io::Error`).

Bob's take: valid. Simple fix.

7. **Rust | Medium | `node/src/main.rs:1469` – `expect` on hardcoded targets.** Parsing `"1d00ffff"` uses `expect`. **Impact:** runtime panic if parsing changes. **Fix:** use `CompactTarget::from_consensus(0x1d00ffff)` or handle the `Result`.

Bob's take: valid. Just switch to `CompactTarget::from_consensus()` instead. Simple fix. Correct fix is correctly deferred with the explicit TODO.

8. **Rust | Medium | `node/src/uncommitted_metadata/mod.rs:23` – Default signature `expect`.** Default metadata parses a hardcoded signature with `expect`. **Impact:** runtime panic if constant changes or parse fails. **Fix:** use pre-validated constant bytes or a fallible constructor.

Bob's take: this is explicitly deferred functionality. Just add a TODO here too.

9. **Rust | Medium | `node/src/committed_metadata/mod.rs:95` – Default pubkey `expect`.** Default metadata parses a hardcoded pubkey with `expect`. **Impact:** runtime panic in default construction. **Fix:** use pre-validated constant bytes or a fallible constructor.

Bob's take: same as above, just add a TODO here.

10. **Rust | Medium | `node/src/utils/mod.rs:84` – `create_test_bead` in production module.** Public function uses test-only constants and `expect`. **Impact:** accidental prod usage with panic paths. **Fix:** move under `#[cfg(test)]` or into `utils::test_utils`.

Bob's take: valid. This should be part of the test framework under `#[cfg(test)]`

11. **Architecture | Medium | `node/src/main.rs:607` – Orphan persistence inferred via slice.** Resolved orphans are inferred from `beads[bead_id+1..]`, coupling main to braid internals. **Impact:** misidentified resolved orphans and fragile persistence logic. **Fix:** have braid return an explicit `resolved_orphans` set.

Bob's take: same as #3.

12. **Architecture | Medium | `node/src/db/db_handlers.rs:599` – DB insert failures not propagated.** Insert errors are logged and dropped. **Impact:** in-memory braid can diverge from DB silently. **Fix:** propagate errors to sync logic and retry/backpressure or fail the batch.

Bob's take: valid. Return `Err` from the handler and have the caller apply backpressure or retry or crash loudly. (e.g. pause IBD if the error is transient). This is a correctness issue. We can't allow the db to diverge from the in-memory representation under any circumstances. It becomes an immediate consensus failure.

13. **Architecture | Medium | `node/src/main.rs:980` – IBD orchestration split across main and PeerManager.** Sync logic lives in main while state lives in PeerManager. **Impact:** tighter coupling and harder testing of IBD flow. **Fix:** reintroduce an IBD manager/state machine with explicit APIs.

Bob's take: This is probably the biggest issue with this PR. PeerManager holds IBD‑specific state (cached bead queues, inflight request tracking, offsets, reset logic), while main.rs drives the actual IBD flow: initiating IBD, validating responses, computing next batch offsets, sending next requests, and handling error paths (disconnect, reset, retry). This means you can't reason about the IBD lifecycle in a single place, and it makes unit testing almost impossible (requiring a full swarm event loop). This has implicit coupling: main has to know how PeerManager's internal IBD fields behave and what resets are required. (more below)

14. **Cryptography | Low | `node/src/main.rs:1469` (and metadata defaults) – Placeholder pubkey/signature/targets.** Mining broadcast still uses hardcoded pubkey/signature and fixed targets. **Impact:** not spec-compliant; unsafe to treat as production-ready. **Fix:** wire to configured miner pubkey, real signatures, and targets derived from difficulty.

Bob's take: explicit future work. Add TODO's where appropriate so we find the right sites when this is implemented in a follow-on PR.

I think I may have made the suggestion to deprecate `ibd_manager` and I think it was a mistake (sorry) and the AI is right on this. Here's some concrete suggestions:
1. IBDManager should own only paging/requests/retries/inflight state. It should not parse/validate/insert beads. Add systematic unit tests for IBDManager. This will remove most of the issues around requiring the entire Swarm to test and make it easier to review and merge.
2. Create a new `ingest_beads()` function that handles validation, braid extension, orphan resolution, DB insert, and scoring. It becomes a one-shot, if it doesn't error, memory and DB are consistent and we're in consensus. If it does, it needs either retry logic or a loud failure and a crash. The application cannot reliably continue.
3. Delete `GetAllBeads` and use only `GetBeadsAfter` with paging. Delete `InsertBeadSequentially` and use only `InsertBeadsBatch` that takes a vector.
4. `extend()` should return the inserted beads.
5. Centralize peer scoring inside `ingest_beads()`. Right now it's split across floodsub and IBD handlers. Compute invalid/added counts once in the ingest path and update the score there.
6. Do DB and scoring work outside the Braid lock.
